compiler/rustc_borrowck/src/diagnostics/mod.rs

```diff
@@ -1133,8 +1132,45 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                                 place_name, partially_str, loop_message
                             ),
                         );
+                        let ty = moved_place.ty(self.body, self.infcx.tcx).ty;
+                        if let ty::Adt(def, ..) = ty.kind()
+                            && Some(def.did()) == self.infcx.tcx.lang_items().pin_type()
+                        {
+                            err.span_suggestion_verbose(
+                                fn_call_span.shrink_to_lo(),
+                                "consider reborrowing the `Pin` instead of moving it",
+                                "as_mut().".to_string(),
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                        if let Some(clone_trait) = tcx.lang_items().clone_trait() {
+                            // We can't use `predicate_may_hold` or `can_eq` without ICEs in
```

Comment:
There should be no type or const variables here -- you can probably erase the lifetimes of the type before calling `predicate_may_hold` or `can_eq` on a fresh inference context here...

Reply:
It still calls InferCtxt::unwrap_region_constraints, which is what causes the ICE.

Reply:
I mean on a *fresh* inference context, not the one you currently have during borrowck.